src/attr.rs

```diff
@@ -15,51 +15,60 @@ pub enum AttrValue<'string> {
     String(&'string str),
-    /// The attribute is set to a non-UTF-8 string.
+    /// The attribute is set to a string that might not be [valid UTF-8](prim@str).
     Bytes(&'string [u8]),
     /// The attribute is not specified.
     Unspecified,
 }
 
-impl<'string> AttrValue<'string> {
-    /// Returns the state of an attribute by inspecting its [value](crate::Repository::get_attr)
-    /// by a [string](prim@str).
-    ///
-    /// As [`str`](prim@str) is guaranteed to contain only valid UTF-8, this function never returns
-    /// [`AttrValue::Bytes`].
-    pub fn from_string(value: Option<&'string str>) -> Self {
-        match unsafe { raw::git_attr_value(value.map_or(ptr::null(), |v| v.as_ptr().cast())) } {
+macro_rules! from_value {
+    ($value:expr => $string:expr) => {
+        match unsafe { raw::git_attr_value($value.map_or(ptr::null(), |v| v.as_ptr().cast())) } {
             raw::GIT_ATTR_VALUE_TRUE => Self::True,
             raw::GIT_ATTR_VALUE_FALSE => Self::False,
-            raw::GIT_ATTR_VALUE_STRING => Self::String(value.unwrap()),
+            raw::GIT_ATTR_VALUE_STRING => $string,
             raw::GIT_ATTR_VALUE_UNSPECIFIED => Self::Unspecified,
             _ => unreachable!(),
         }
+    };
+}
+
+impl<'string> AttrValue<'string> {
+    /// Returns the state of an attribute by inspecting its [value](crate::Repository::get_attr)
+    /// by a [string](prim@str).
+    ///
+    /// This function always returns [`AttrValue::String`] and never returns [`AttrValue::Bytes`]
+    /// when the attribute is set to a string.
+    pub fn from_string(value: Option<&'string str>) -> Self {
+        from_value!(value => Self::String(value.unwrap()))
     }
 
     /// Returns the state of an attribute by inspecting its [value](crate::Repository::get_attr_bytes)
     /// by a [byte](u8) [slice].
+    ///
+    /// This function will perform UTF-8 validation when the attribute is set to a string, returns
+    /// [`AttrValue::String`] if it's valid UTF-8 and [`AttrValue::Bytes`] otherwise.
     pub fn from_bytes(value: Option<&'string [u8]>) -> Self {
-        match unsafe { raw::git_attr_value(value.map_or(ptr::null(), |v| v.as_ptr().cast())) } {
-            raw::GIT_ATTR_VALUE_TRUE => Self::True,
-            raw::GIT_ATTR_VALUE_FALSE => Self::False,
-            raw::GIT_ATTR_VALUE_STRING => {
-                let value = value.unwrap();
-                if let Ok(string) = str::from_utf8(value) {
-                    Self::String(string)
-                } else {
-                    Self::Bytes(value)
-                }
+        from_value!(value => {
+            let value = value.unwrap();
+            if let Ok(string) = str::from_utf8(value) {
+                Self::String(string)
+            } else {
+                Self::Bytes(value)
             }
-            raw::GIT_ATTR_VALUE_UNSPECIFIED => Self::Unspecified,
-            _ => unreachable!(),
-        }
+        })
```

Comment:
Could this method perhaps call `always_bytes` instead of reusing a macro to call the same C function?

Reply:
Good idea! :+1: